x-pack/plugins/rollup/server/routes/api/index_patterns.js

```diff
@@ -3,86 +3,108 @@
-          const indexCapabilities = getCapabilitiesForIndexName(allCapabilities, indexName);
-          const jobs = indexCapabilities && Object.keys(indexCapabilities);
-          const fieldsFromFieldCapsByName = jobs && indexCapabilities[jobs[0]].fields;
+      const rollupIndex = params.rollup_index;
+      const callWithRequest = callWithRequestFactory(server, request);
 
-          // Keep meta fields
-          JSON.parse(metaFields).forEach(field => fieldsFromIndex[field] && rollupFields.push(fieldsFromIndex[field]));
+      try {
+        const rollupFields = [];
+        const rollupFieldNames = [];
+        const fieldsFromFieldCapsApi = indexBy(await getFieldCapabilities(callWithRequest, [rollupIndex], metaFields), 'name');
+        const rollupIndexCapabilities = getCapabilitiesForRollupIndices(await callWithRequest('rollup.capabilitiesByRollupIndex', {
+          indexPattern: rollupIndex
+        }))[rollupIndex].aggs;
 
-          // Map rollup capabilities
-          Object.keys(fieldsFromFieldCapsByName).forEach(capField => {
-            const capabilities = fieldsFromFieldCapsByName[capField].reduce((aggs, cap) => aggs.push(cap.agg) && aggs, []);
-            const defaultField = {
-              name: capField,
-              searchable: true,
-              aggregatable: true,
-              readFromDocValues: true,
-            };
+        // Keep meta fields
+        metaFields.forEach(field => fieldsFromFieldCapsApi[field] && rollupFields.push(fieldsFromFieldCapsApi[field]));
 
-            if(capabilities.includes('date_histogram')) {
-              rollupFields.push({
-                ...fieldsFromIndex[`${capField}.date_histogram.timestamp`],
-                ...defaultField,
-              });
-            } else if(capabilities.includes('histogram')) {
-              rollupFields.push({
-                ...fieldsFromIndex[`${capField}.histogram.value`],
-                ...defaultField,
-              });
-            } else if(capabilities.includes('terms')) {
-              rollupFields.push({
-                ...fieldsFromIndex[`${capField}.terms.value`],
-                ...defaultField,
-              });
-            } else {
-              rollupFields.push({
-                ...defaultField,
-                type: 'number',
-                searchable: false,
-              });
-            }
-          });
+        // Merge rollup capabilities information with field information
+        Object.keys(rollupIndexCapabilities).forEach(agg => {
+          const fields = Object.keys(rollupIndexCapabilities[agg]);
+          const defaultField = {
+            name: null,
+            searchable: true,
+            aggregatable: true,
+            readFromDocValues: true,
+          };
 
-          return reply({
-            fields: rollupFields,
-          });
-        } catch(err) {
-          if (isEsError(err)) {
-            return reply(wrapEsError(err));
+          switch(agg) {
```

Comment:
Please check my logic on this, but it looks like we're duplicating some code here? Also, maybe we could add a comment on why we're caching the filtered fields?

```js
          const regularFields = fields.filter(field => !rollupFieldNames.includes(field));

          // Cache this for later for some reason?
          rollupFieldNames.push(...regularFields);

          switch(agg) {
            case 'date_histogram':
              const fieldsWithRollupCapabilities = map(field => {
                return {
                  ...fieldsFromFieldCapsApi[`${field}.${agg}.timestamp`],
                  ...defaultField,
                  name: field,
                };
              });

              rollupFields.push(...fieldsWithRollupCapabilities);
              break;
            default:
              const fieldsWithRollupCapabilities = map(field => {
                return {
                  ...fieldsFromFieldCapsApi[`${field}.${agg}.value`],
                  ...defaultField,
                  name: field,
                };
              });

              rollupFields.push(...fieldsWithRollupCapabilities);
              break;
          }
```

Reply:
I de-duped the code here, got rid of the switch altogether. the only thing that actually needed to be different is `${field}.${agg}.timestamp` vs `${field}.${agg}.value` so thanks for catching that.

the rationale behind this code is:
- We get rollup aggregation capabilities information back in this kind of format. Note that `bytes` belongs to more than one aggregation, and we don't have any information as what type fields are (numeric, keyword, etc):
```
{
  logstash_rollup: {
    aggs: {
        avg: { bytes: {agg: "avg"}, memory: {agg: "avg"}},
        min: { bytes: {agg: "min"} }
        ...
    }
  },
  ...
}
```
- We get fields information back via regular field capabilities (`getFieldCapabilities()`) that tells us type, but they are keyed with internal rollup field names:
```
{
  fields: {
    "bytes.min.value": {
      name: "bytes.min.value",
      type: "number",
      searchable: true,
      aggregatable: true,
      readFromDocValues: true
    },
    ...
  }
}
```
- Logic in this endpoint looks through each aggregation, and for each field in the aggregation, pulls the relevant information from field capabilities by using the key `${field}.${agg}.value` (or `${field}.${agg}.date_histogram` for date histogram aggs)
- Since a field can have multiple aggregations, if we don't cache the rollup fields we have already collected, we end up with duplicate field information